tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp

```diff
@@ -29,22 +32,125 @@ bool operator==(int x, const Cat& c) {
+        assert(find(begin(sc), end(sc), ValueType{5}) == end(sc));
+
+        if constexpr (sizeof(ElementType) <= sizeof(ValueType)) {
+            assert(find(begin(sc), end(sc), ValueType{max_val}) == begin(sc) + 12);
+            assert(count(begin(sc), end(sc), ValueType{max_val}) == 1);
+
+            if constexpr (sizeof(ElementType) < sizeof(ValueType)) {
+                assert(find(begin(sc), end(sc), ValueType{ValueType{max_val} + 1}) == end(sc));
+                assert(find(begin(sc), end(sc), ValueType{umax_val}) == end(sc));
+
+                assert(count(begin(sc), end(sc), ValueType{ValueType{max_val} + 1}) == 0);
+                assert(count(begin(sc), end(sc), ValueType{umax_val}) == 0);
+            }
+        }
+    } else {
+        constexpr ElementType max_val = numeric_limits<ElementType>::max();
+        constexpr ValueType max_vt    = numeric_limits<ValueType>::max();
+
+        const ElementType uc[] = {0, 1, 1, 2, max_val - 2, max_val - 1, max_val};
+
+        assert(find(begin(uc), end(uc), ValueType{0}) == begin(uc));
+        assert(find(begin(uc), end(uc), ValueType{2}) == begin(uc) + 3);
+        assert(find(begin(uc), end(uc), ValueType{6}) == end(uc));
+
+        if constexpr (max_val <= max_vt) {
+            assert(find(begin(uc), end(uc), ValueType{max_val - 3}) == end(uc));
+            assert(find(begin(uc), end(uc), ValueType{max_val - 2}) == begin(uc) + 4);
+            assert(find(begin(uc), end(uc), ValueType{max_val}) == begin(uc) + 6);
+            if constexpr (max_val < max_vt) {
+                assert(find(begin(uc), end(uc), ValueType{ValueType{max_val} + 1}) == end(uc));
+                if constexpr (sizeof(ElementType) < sizeof(ValueType)) {
+                    assert(find(begin(uc), end(uc), max_vt) == end(uc));
+                }
+            }
+        }
+    }
+}
+
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif // __clang__
+#pragma warning(pop)
+
+template <class ElementType>
+void test_limit_check_elements() {
+    test_limit_check_elements_impl<ElementType, signed char>();
+    test_limit_check_elements_impl<ElementType, short>();
+    test_limit_check_elements_impl<ElementType, int>();
+    test_limit_check_elements_impl<ElementType, long>();
+    test_limit_check_elements_impl<ElementType, long long>();
+
+    test_limit_check_elements_impl<ElementType, unsigned char>();
+    test_limit_check_elements_impl<ElementType, unsigned short>();
+    test_limit_check_elements_impl<ElementType, unsigned int>();
+    test_limit_check_elements_impl<ElementType, unsigned long>();
+    test_limit_check_elements_impl<ElementType, unsigned long long>();
+}
+
 int main() {
     { // DevDiv-316853 "<algorithm>: find()'s memchr() optimization is incorrect"
+
+
         vector<signed char> v;
         v.push_back(22);
         v.push_back(33);
         v.push_back(-1);
         v.push_back(44);
+        v.push_back(44);
         v.push_back(55);
 
+#ifdef __cpp_lib_concepts

```

Comment:
why are these guarded by concepts?

Reply:
`std::vector` iterators are not detected as contiguous without concepts.
We might look into it, and expand optimization for pre-C++20, but even if we do, i prefer it i a next PR.